py/desispec/io/meta.py

```diff
@@ -17,14 +17,65 @@
 from desiutil.log import get_logger
 from .util import healpix_subdirectory, checkgzip
 
+_desi_root_readonly=None
+def get_desi_root_readonly():
+    """
+    Returns $DESI_ROOT_READONLY if set and path exists, otherwise $DESI_ROOT.
+
+    Caches answer upon first call, i.e. setting $DESI_ROOT_READONLY to a
+    different value part way through running will use previously cached value.
+    This prevents it from re-checking a non-existent path N>>1 times.
+    """
+    global _desi_root_readonly
+    log = get_logger()
+    if _desi_root_readonly is not None:
+        log.debug('Using cached _desi_root_readonly=%s', _desi_root_readonly)
+        return _desi_root_readonly
+    elif 'DESI_ROOT_READONLY' in os.environ:
+        if os.path.exists(os.environ['DESI_ROOT_READONLY']):
+            _desi_root_readonly = os.environ['DESI_ROOT_READONLY']
+            log.debug("Using $DESI_ROOT_READONLY=%s", _desi_root_readonly)
```

Comment:
Formatting+argument will only evaluate the string inside the logger call if the log level is high enough that it knows it will will print the string, while f-string formatting will always happen before being passed into the logger call.  Thus format+argument(s) is preferred for debug-level calls that normally aren't printed.  We've had cases in the past were significant time was spent on f-string formatting of debug messages that never got printed deep within loops, and switching to format+arguments was faster.  Some people strictly use format+args for any log messages for this reason; I tend to to it only for debug messages because I overall prefer the clarity of f-string formatting and performance-wise it doesn't matter if the message is going to be printed either way.

Reply:
Thank you for the explanation, that makes perfect sense.

Reply:
as a paranoia check I also tried running the unit tests with DESI_LOGLEVEL=debug to ensure that I didn't have a typo in my debug log statements...